src/middlewared/middlewared/alert/source/ntp_status.py

```diff
@@ -0,0 +1,162 @@
+FIELDS_CNT = len(FIELDS)
+
+
+def since_boot():
+    return time.time() - psutil.boot_time()
+
+
+def reachability(byte):
+    return '{0:08b}'.format(byte)
+
+
+def parse_peers(output):
+    peers = []
+
+    try:
+        remote = ''
+        select = ''
+        for line in output.rpartition('=====')[2].strip().splitlines():
+            """ """
+            peer = line.strip().split()
+            peer_cnt = len(peer)
+
+            """Extract peer select status and peer address"""
+            if peer_cnt in (1, FIELDS_CNT):
+                match = SELECT_STATUS_RE.match(peer[0])
+                select = match.group(0) if match else ''
+                remote = peer[0].lstrip(select)
+                if peer_cnt == 1:
+                    continue
+                peer[0] = remote
+            elif peer_cnt == FIELDS_CNT-1 and remote:
+                """Continuation line"""
+                peer.insert(0, remote)
+                remote = ''
+            else:
+                raise Exception("invalid row: '{}'".format(line))
+
+            info = dict(zip(FIELDS, peer))
+
+            info['select'] = SELECT_STATUS.get(select, '')['message']
+            info['type'] = TYPE.get(info['type'], '-')
+            info['reach'] = int(info['reach'], 8)
+
+            if info['refid'] == '.INIT.':
+                if since_boot() < STARTUP_PERIOD:
+                    break
+                else:
+                    raise Exception("still initializing after {} seconds from boot".format(since_boot()))
+
+            """Skip discarded peers """
+            if info['select'] not in ('sel_candidate', 'sel_backup', 'sel_sys.peer', 'sel_pps.peer'):
+                continue
+
+            peers.append(info)
+            remote = ''
+            select = ''
+
+    except Exception as e:
+            logger.warning("NTP status: {}".format(str(e)))
+            return []
+
+    return peers
+
+
+class NTPStatusAlertSource(AlertSource):
+    level = AlertLevel.CRITICAL
+    title = "NTP status"
+
+    schedule = IntervalSchedule(timedelta(minutes=1))
+
+    async def _get_peers(self):
+        if os.path.exists('/var/tmp/ntp/query'):
+            with open('/var/tmp/ntp/query', encoding='utf-8') as query:
+                peers = query.read()
+        else:
+            peers = (await run(['/usr/bin/ntpq', '-pwn'], encoding='utf8')).stdout
+
+        return parse_peers(peers)
+
+    async def check(self):
+        alerts = []
+
+        info = next((peer for peer in await self._get_peers() if peer.get('select', '-') in ('sel_sys.peer')), None)
+        if info:
+            """ Success of the last 4 probes """
+            if info['reach'] < 255 and info['reach'] & 0x0F != 0x0F:
+                alerts.append(Alert(
+                        title="NTP status: %(times)d out of 4 last probes failed",
+                        args={'times': reachability(info['reach']).count('0')},
```

Comment:
We can add a constant `key` here, does it really matter how many times it failed?

Reply:
Good question. This is a live number that shows in a bit vector last 8 successful and failed probes. Not sure what is the best option here. Now in the worse case scenario we'll get 4 warnings. If we put constant key, wouldn't that freeze the status of the first probe failure and all subsequent failures will become unnoticed?

Reply:
@b-a-t maybe special key for 4 probes?